### PR TITLE
Implement `.by`

### LINKED ISSUE #1051: Implement `.by`
This makes code so much nicer! 😄

### PR DESCRIPTION
**Pull request by mgirlich**
Closes #1051.

### CODE REVIEW

**Comment by mgirlich** [comment]
@hadley Will you be able to review the multi join PR in the next weeks? Together with the support of the new dplyr 1.1.0 features this would be quite a nice release for dbplyr.

**Comment by hadley** [comment]
Yeah, for sure. We're planning on submitting dplyr to CRAN on Jan 27, so maybe we should plan for a dbplyr submission on Jan 28? (Unless you think we can break the dependency on dev dplyr, in which case we could release earlier).

File: R/verb-filter.R [hunk 1]
```diff
@@ -15,11 +15,19 @@
 #' db %>% filter(is.na(x)) %>% show_query()
 # registered onLoad
 #' @importFrom dplyr filter
-filter.tbl_lazy <- function(.data, ..., .preserve = FALSE) {
+filter.tbl_lazy <- function(.data, ..., .by = NULL, .preserve = FALSE) {
   if (!identical(.preserve, FALSE)) {
     cli_abort("{.arg .preserve} is not supported on database backends")
   }
   check_filter(...)
+  by <- compute_by({{ .by }},
```
Comment (hadley) [inline R/verb-filter.R:23]: Indenting style? You might want to try to http://github.com/lionel-/codegrip/ do toggle between tidyverse styles.

File: R/verb-slice.R [hunk 2]
```diff
@@ -48,13 +49,13 @@ NULL
 
 #' @importFrom dplyr slice
 #' @export
-slice.tbl_lazy <- function(.data, ...) {
+  slice.tbl_lazy <- function(.data, ...) {
```
Comment (hadley) [inline R/verb-slice.R:52]:
```suggestion
slice.tbl_lazy <- function(.data, ...) {
```

File: R/verb-slice.R [hunk 3]
```diff
@@ -110,10 +111,20 @@ slice_sample.tbl_lazy <- function(.data, ..., n, prop, weight_by = NULL, replace
     cli_abort("Sampling with replacement is not supported on database backends")
   }
 
-  slice_by(.data, !!sql_random(remote_con(.data)), size, with_ties = FALSE)
+  slice_by(.data, !!sql_random(remote_con(.data)), size, {{ by }}, with_ties = FALSE)
 }
 
-slice_by <- function(.data, order_by, size, with_ties = FALSE) {
+slice_by <- function(.data, order_by, size, .by, with_ties = FALSE) {
+  # browser()
```
Comment (hadley) [inline R/verb-slice.R:118]: Can remove now?

File: tests/testthat/_snaps/verb-slice.md [hunk 4]
```diff
@@ -112,3 +112,29 @@
       Error in `slice_sample()`:
       ! `prop` must be a non-missing positive number.
 
+# catches `.by` with grouped-df
```
Comment (hadley) [inline tests/testthat/_snaps/verb-slice.md:115]: I'm not sure I'd test this in quite so many places

**Review by hadley — APPROVED** [review]

**Pull request merged**